File: JavaScript/5-partial-ext.js
'use strict';

const partial = (fn, ...args) => (...rest) => fn(...args, ...rest);

// Usage

const sum4 = (a, b, c, d) => (a + b + c + d);

const f11 = partial(sum4, 1);
const f12 = partial(f11, 2);
const f13 = partial(f12, 3);
const y1 = f13(4);
console.log(y1);

const f21 = partial(sum4, 1, 2);
const f22 = partial(f21, 3);
const y2 = f22(4);
console.log(y2);

File: Python/5-partial-ext.py

def partial(fn, *args):
    return lambda *rest: fn(*(args + rest))


def summary(a, b, c, d):
    return a + b + c + d


if __name__ == '__main__':
    f1 = partial(summary, 1)
    f2 = partial(f1, 2)
    f3 = partial(f2, 3)
    print(f3(4))

    f11 = partial(summary, 1, 2)
    f12 = partial(f11, 3)
    print(f12(4))
